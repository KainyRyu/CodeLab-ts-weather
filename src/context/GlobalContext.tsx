import React, { createContext, useContext, useEffect, useState } from 'react';
import { setLocalStorage } from 'lib/helper';

interface Props {
  children: JSX.Element | JSX.Element[];
}
interface Theme {
  icon: string;
  theme: { color: string; backgroundColor: string };
  imgUrl?: string;
}
interface GlobalValue {
  cities: string[];
  selectedCity: string;
  weatherTheme: Theme;
  setWeatherTheme: React.Dispatch<React.SetStateAction<Theme>>;
  setSelectedCity: React.Dispatch<React.SetStateAction<string>>;
  setCities: React.Dispatch<React.SetStateAction<string[]>>;
  handleCityClick: (value: string) => void;
}

const GlobalContext = createContext<GlobalValue>({
  cities: [],
  setCities: () => {},
  selectedCity: '',
  weatherTheme: { icon: '', theme: { color: '', backgroundColor: '' }, imgUrl: '' },
  setWeatherTheme: () => {},
  setSelectedCity: () => '',
  handleCityClick: () => {},
});

function GlobalProvider({ children }: Props) {
  const cityListInit = ['london', 'berlin', 'seoul', 'paris', 'prague'];
  const [cities, setCities] = useState(cityListInit);
  const [selectedCity, setSelectedCity] = useState('london');
  const [weatherTheme, setWeatherTheme] = useState<Theme>({
    icon: 'â˜€',
    theme: { color: '#FFA820', backgroundColor: '#FFF8EE' },
    imgUrl: '',
  });

  useEffect(() => {
    setLocalStorage({ key: 'cities', value: cities });
  }, [cities]);

  const handleCityClick = (value: string) => {
    setSelectedCity(value);
  };

  return (
    <GlobalContext.Provider
      value={{
        cities,
        selectedCity,
        weatherTheme,
        setWeatherTheme,
        setSelectedCity,
        setCities,
        handleCityClick,
      }}>
      {children}
    </GlobalContext.Provider>
  );
}

const useGlobalContext = () => useContext(GlobalContext);

export { useGlobalContext, GlobalProvider };
