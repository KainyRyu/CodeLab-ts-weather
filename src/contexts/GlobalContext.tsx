import { useEffect, useContext, createContext } from 'react';
import { CITIES } from 'data';
import { getDataFromSessionStorage, setSessionStorage } from 'helper';

interface Props {
  children: JSX.Element | JSX.Element[];
}

const GlobalContext = createContext({});

const GlobalProvider = ({ children }: Props) => {
  useEffect(() => {
    setSessionStorage('locationList', CITIES);
  }, []);

  const myList: string | null = getDataFromSessionStorage('locationList');

  const addToList = () => {};

  return <GlobalContext.Provider value={{ myList, addToList }}>{children}</GlobalContext.Provider>;
};

const useGlobalContext = () => useContext(GlobalContext);

export { GlobalContext as default, useGlobalContext, GlobalProvider };
