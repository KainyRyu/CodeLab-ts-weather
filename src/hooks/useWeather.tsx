import { useQuery } from 'react-query';
import axios from 'axios';

const APIKEY: string = 'bd5de54bb8eed93db96a311f088b9415';

const getWeatherBycityname = async (cityname: string) => {
  const { data } = await axios.get(
    `http://api.openweathermap.org/data/2.5/weather?q=${cityname}&appid=${APIKEY}&units=metric`,
  );
  return data;
};

export default function useWeather(cityname: string) {
  return useQuery(['post', cityname], () => getWeatherBycityname(cityname));
}

const getWeatherByCords = async (lat: number, lon: number) => {
  const { data } = await axios.get(
    `http://api.openweathermap.org/data/2.5/weather?lat=${lat}&lon=${lon}&appid=${APIKEY}&units=metric`,
  );
  return data;
};

export function useWeatherByCoords(lat: number, lon: number) {
  return useQuery(['location'], () => getWeatherByCords(lat, lon));
}
