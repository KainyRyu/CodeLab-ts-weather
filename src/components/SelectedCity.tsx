import React, { useEffect } from 'react';
import Styled from 'styled-components';
import { useGlobalContext } from 'context/GlobalContext';
import useWeather from 'hooks/useWeather';
import { weatherThemes } from 'lib/config';
import Img404 from 'assets/404.png';

interface Props {
  readonly selectedCity: string;
}

interface Error {
  readonly config: {
    transitional: {};
    transformRequest: [];
    transformResponse: [];
    timeout: number;
  };
  readonly isAxiosError: boolean;
  readonly request: {};
  readonly response: {
    data: { cod: string; message: string };
    status: number;
    statusText: string;
    headers: {};
  };
  readonly toJSON: () => void;
}

const WeatherImg = Styled.img`
  width: 100px;
  height: 100px
`;

const Li = Styled.span`
  display: flex;
  justify-content: space-between;
`;

const LostImg = Styled.img`
  width: 400px;
  height: auto;
`;

export default function SelectedCity({ selectedCity }: Props) {
  const { weatherTheme, setWeatherTheme } = useGlobalContext();

  const { isError, isLoading, data, error } = useWeather(selectedCity);

  const theme = (weatherCode: number) => {
    switch (`${weatherCode}`[0]) {
      case '2':
        return weatherThemes.thunderstorm;
      case '3':
        return weatherThemes.drizzle;
      case '5':
        return weatherThemes.rain;
      case '6':
        return weatherThemes.snow;
      case '7':
        return weatherThemes.atmosphere;
      case '8':
        return weatherCode === 800 ? weatherThemes.clear : weatherThemes.cloud;
    }
  };

  useEffect(() => {
    !isLoading && data && setWeatherTheme(theme(data?.weather[0].id)!);
  }, [isLoading, data, setWeatherTheme]);

  return isLoading ? (
    <div>loading...</div>
  ) : isError ? (
    <div>
      <h3>
        {selectedCity} : {(error as Error).response.data.message}
      </h3>
      <LostImg src={Img404} alt="404" />
      {/* <div>
        {(error as Error).response.data.cod}: {(error as Error).response.data.message}
      </div> */}
    </div>
  ) : (
    <div>
      <h1>{data.name}</h1>
      <WeatherImg src={weatherTheme.imgUrl} alt="weather" />
      {/* weather api's icon
      <WeatherImg
        src={` http://openweathermap.org/img/wn/${data.weather[0].icon}@2x.png`}
        alt="weather"
      /> */}
      <h3>{data.weather[0].main}</h3>
      <Li>
        <span>Temprature</span>
        <span>{data.main.temp}℃</span>
      </Li>
      <Li>
        <span>Feels like</span>
        <span>{data.main.feels_like}℃</span>
      </Li>
      <Li>
        <span>Humidity</span>
        <span>{data.main.humidity}%</span>
      </Li>
      <Li>
        <span>wind</span>
        <span>{data.wind.speed}m/s</span>
      </Li>
    </div>
  );
}
