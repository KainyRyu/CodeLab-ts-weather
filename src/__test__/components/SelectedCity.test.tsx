import { render, waitFor } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import SelectedCity from 'components/SelectedCity';
import { GlobalProvider } from 'context/GlobalContext';
import useWeather from 'hooks/useWeather';
import React from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';

interface Props {
  children: JSX.Element | JSX.Element[];
}

describe('<SelectedCity />', () => {
  it('renders component correctly', () => {
    const { container } = render(<SelectedCity selectedCity="seoul" />);

    expect(container).toMatchSnapshot();
  });

  it("display seoul's weather", async () => {
    const wrapper = ({ children }: Props) => {
      const queryClient = new QueryClient({
        defaultOptions: {
          queries: {
            retry: false,
          }
        }
      });
      return (
        <QueryClientProvider client={queryClient}>
          <GlobalProvider>
            <SelectedCity selectedCity="seoul" />
          </GlobalProvider>
        </QueryClientProvider>
      );
    };

    const { result, waitFor } = renderHook(() => useWeather('seoul'), { wrapper });
    await waitFor(() => result.current.isSuccess);
    expect(result.current.data.name).toEqual('Seoul');
  });

  // it('displays not found message', async () => {
  //   const wrapper = ({ children }: Props) => {
  //     const queryClient = new QueryClient({
  //       defaultOptions: {
  //         queries: {
  //           retry: false,
  //         },
  //       },
  //     });
  //     return (
  //       <QueryClientProvider client={queryClient}>
  //         <GlobalProvider>
  //           <SelectedCity selectedCity="abc" />
  //         </GlobalProvider>
  //       </QueryClientProvider>
  //     );
  //   };

  //   const { result, waitFor } = renderHook(() => useWeather('abc'), { wrapper });
  //   await waitFor(() => result.current.isError);
  //   expect(result).toEqual('Seoul');
  // });
});
