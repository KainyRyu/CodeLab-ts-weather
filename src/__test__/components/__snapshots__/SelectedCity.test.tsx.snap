// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectedCity /> renders component correctly 1`] = `
<div>
  <div />
</div>
`;
